src/quic/node_quic_session.cc

```diff
@@ -3222,9 +3231,29 @@ int QuicSession::OnStatelessReset(
   return 0;
 }
 
+BaseObjectPtr<QLogStream> QuicSession::qlog_stream() {
+  if (!qlog_stream_) {
+    CHECK(qlog_stream_ = QLogStream::Create(env()));
+    listener_->OnQLog(qlog_stream_.get());
+  }
+  return qlog_stream_;
+}
+
 void QuicSession::OnQlogWrite(void* user_data, const void* data, size_t len) {
   QuicSession* session = static_cast<QuicSession*>(user_data);
-  session->listener()->OnQLog(reinterpret_cast<const uint8_t*>(data), len);
+  Environment* env = session->env();
+  if (env->can_call_into_js()) {
+    // If we're able to call into JavaScript, dispatch the chunk immediately.
+    session->qlog_stream()->Emit(reinterpret_cast<const uint8_t*>(data), len);
+  } else {
+    BaseObjectPtr<QLogStream> ptr = session->qlog_stream();
+    std::vector<uint8_t> buffer(len);
+    memcpy(buffer.data(), data, len);
+    env->SetImmediate([ptr = std::move(ptr),
+                       buffer = std::move(buffer)](Environment*) {
+      ptr->Emit(buffer.data(), buffer.size());
+    });
+  }
```

Comment:
I’m not sure that I understand this `if ... else` fully – isn’t `ptr->Emit()` still calling into JS, even if it’s inside a `SetImmediate()`?

Reply:
Per our slack conversation, I changed it so that Emits always defer to immediate. This should be safe because there's nothing timing critical about the qlog output so long as we get the complete stream.